<?xml version="1.0"?>
<robot name="my_bot" xmlns:xacro="http://www.ros.org/wiki/xacro">
	
  <!-- Base link -->
  <link name="base_link">
    <visual>
      <geometry><box size="0.24 0.21 0.14"/></geometry>
      <material name="blue"><color rgba="0 0 0.8 1.0"/></material>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </visual>
    <collision>
      <geometry><box size="0.24 0.21 0.14"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <mass value="0.2"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.02"/>
    </inertial>
  </link>

  <!-- Gazebo base link properties -->
  <gazebo reference="base_link">
    <material>Gazebo/Blue</material>
    <mu1>0.8</mu1>
    <mu2>0.8</mu2>
  </gazebo>

  <!-- Wheels -->
  <link name="left_wheel_back">
    <visual>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <material name="wheel_black"><color rgba="0.15 0.15 0.15 1.0"/></material>
    </visual>
    <collision>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0"/>
      <mass value="0.02"/>  
      <inertia ixx="0.00013" ixy="0" ixz="0" iyy="0.00023" iyz="0" izz="0.00013"/>
    </inertial>
  </link>

  <gazebo reference="left_wheel_back">
    <material>Gazebo/Black</material>
    <mu1>100.0</mu1>
    <mu2>1000.0</mu2>
  </gazebo>

  <link name="right_wheel_back">
    <visual>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <material name="wheel_black"><color rgba="0.15 0.15 0.15 1.0"/></material>
    </visual>
    <collision>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0"/>
      <mass value="0.02"/>
      <inertia ixx="0.00013" ixy="0" ixz="0" iyy="0.00023" iyz="0" izz="0.00013"/>
    </inertial>
  </link>

  <gazebo reference="right_wheel_back">
    <material>Gazebo/Black</material>
    <mu1>100.0</mu1>
    <mu2>1000.0</mu2>
  </gazebo>

  <link name="left_wheel_front">
    <visual>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <material name="wheel_black"><color rgba="0.15 0.15 0.15 1.0"/></material>
    </visual>
    <collision>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0"/>
      <mass value="0.02"/>
      <inertia ixx="0.00013" ixy="0" ixz="0" iyy="0.00023" iyz="0" izz="0.00013"/>
    </inertial>
  </link>

  <gazebo reference="left_wheel_front">
    <material>Gazebo/Black</material>
    <mu1>100.0</mu1>
    <mu2>1000.0</mu2>
  </gazebo>

  <link name="right_wheel_front">
    <visual>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <material name="wheel_black"><color rgba="0.15 0.15 0.15 1.0"/></material>
    </visual>
    <collision>
      <geometry><cylinder length="0.05" radius="0.05"/></geometry>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0"/>
      <mass value="0.02"/>
      <inertia ixx="0.00013" ixy="0" ixz="0" iyy="0.00023" iyz="0" izz="0.00013"/>
    </inertial>
  </link>

  <gazebo reference="right_wheel_front">
    <material>Gazebo/Black</material>
    <mu1>10.00</mu1>
    <mu2>1000.0</mu2>
  </gazebo>

  <!-- Wheel joints -->
  <joint name="left_wheel_back_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel_back"/>
    <origin xyz="-0.12 0.13 -0.07" rpy="-1.5707 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="right_wheel_back_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel_back"/>
    <origin xyz="-0.12 -0.13 -0.07" rpy="1.5707 0 0"/>
    <axis xyz="0 0 -1"/>
  </joint>

  <joint name="left_wheel_front_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel_front"/>
    <origin xyz="0.12 0.13 -0.07" rpy="-1.5707 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <joint name="right_wheel_front_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel_front"/>
    <origin xyz="0.12 -0.13 -0.07" rpy="1.5707 0 0"/>
    <axis xyz="0 0 -1"/>
  </joint>

  <link name="camera_link">
    <visual>
      <geometry><box size="0.02 0.05 0.05"/></geometry>
      <material name="gray"/>
    </visual>
  </link>

  <joint name="camera_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_link"/>
    <origin xyz="0.1 0 0.1" rpy="0 0 0"/> </joint>

  <link name="camera_link_optical"></link>
  <joint name="camera_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}" />
    <parent link="camera_link" />
    <child link="camera_link_optical" />
  </joint>

  <gazebo reference="camera_link">
    <sensor name="rgbd_camera" type="depth">
      <update_rate>30</update_rate>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.1</near>
          <far>100</far>
        </clip>
      </camera>

      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <frame_name>camera_link_optical</frame_name>
        
        <image_topic_name>rgb/image_raw</image_topic_name>
        <camera_info_topic_name>rgb/camera_info</camera_info_topic_name>
        
        <depth_image_topic_name>depth/image_raw</depth_image_topic_name>
        <depth_image_camera_info_topic_name>depth/camera_info</depth_image_camera_info_topic_name>
        <depth_image_format>R_FLOAT32</depth_image_format>
        
        <point_cloud_topic_name>depth/points</point_cloud_topic_name>
      </plugin>
    </sensor>
  </gazebo>
  <!-- IMU -->
  <gazebo reference="base_link">
    <sensor name="imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>50</update_rate>
      <plugin name="imu_plugin" filename="libgazebo_ros_imu_sensor.so">
        <ros><namespace>/</namespace><remapping>~/out:=imu</remapping></ros>
        <frame_name>base_link</frame_name>
      </plugin>
    </sensor>
  <!-- GPS -->
    <sensor name="gps_sensor" type="gps">
      <always_on>true</always_on>
      <update_rate>1.0</update_rate>
      <plugin name="gps_plugin" filename="libgazebo_ros_gps_sensor.so">
        <ros><namespace>/</namespace><remapping>~/out:=gps</remapping></ros>
        <frame_name>base_link</frame_name>
      </plugin>
    </sensor>
  </gazebo>
  
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>$(find my_bot_description)/config/my_bot_controllers.yaml</parameters>
    </plugin>
  </gazebo>

    <xacro:diffbot_ros2_control
      name="diffbot_system"
      prefix=""
      use_mock_hardware="true" />
 


</robot>

